src/Aspire.Hosting/Utils/PasswordGenerator.cs

```diff
@@ -14,55 +15,94 @@ internal static class PasswordGenerator
+        var length = Math.Max(minLength, requiredMinLength);
 
-        ArgumentOutOfRangeException.ThrowIfZero(length);
+        Span<char> chars = stackalloc char[length];
 
-        Debug.Assert(length <= 128, "password too long");
+        // fill the required characters first
+        var currentChars = chars;
+        GenerateRequiredValues(ref currentChars, minLower, LowerCaseChars);
+        GenerateRequiredValues(ref currentChars, minUpper, UpperCaseChars);
+        GenerateRequiredValues(ref currentChars, minNumeric, NumericChars);
+        GenerateRequiredValues(ref currentChars, minSpecial, SpecialChars);
 
-        Span<char> chars = stackalloc char[length];
+        // fill the rest of the password with random characters from all the available choices
+        var choices = GetChoices(lower, upper, numeric, special);
+        RandomNumberGenerator.GetItems(choices, currentChars);
 
-        RandomNumberGenerator.GetItems(LowerCaseChars, chars.Slice(0, lowerCase));
-        RandomNumberGenerator.GetItems(UpperCaseChars, chars.Slice(lowerCase, upperCase));
-        RandomNumberGenerator.GetItems(DigitChars, chars.Slice(lowerCase + upperCase, digit));
-        RandomNumberGenerator.GetItems(SpecialChars, chars.Slice(lowerCase + upperCase + digit, special));
         RandomNumberGenerator.Shuffle(chars);
 
         return new string(chars);
     }
 
-    /// <summary>
-    /// Creates a random string of upper and lower case letters.
-    /// </summary>
-    public static string GenerateRandomLettersValue(int length)
+    private static void CheckMinZeroWhenDisabled(
+        bool enabled,
+        int minValue,
+        [CallerArgumentExpression(nameof(enabled))] string? enabledParamName = null,
+        [CallerArgumentExpression(nameof(minValue))] string? minValueParamName = null)
     {
-        ArgumentOutOfRangeException.ThrowIfNegative(length);
-        ArgumentOutOfRangeException.ThrowIfGreaterThan(length, 128);
+        if (!enabled && minValue > 0)
+        {
+            ThrowArgumentException();
+        }
 
-        Span<char> chars = stackalloc char[length];
+        void ThrowArgumentException() => throw new ArgumentException($"'{minValueParamName}' must be 0 if '{enabledParamName}' is disabled.");
+    }
 
-        RandomNumberGenerator.GetItems(LettersChars, chars);
+    private static void GenerateRequiredValues(ref Span<char> destination, int minValues, string choices)
+    {
+        Debug.Assert(destination.Length >= minValues);
 
-        return new string(chars);
+        if (minValues > 0)
+        {
+            RandomNumberGenerator.GetItems(choices, destination.Slice(0, minValues));
+            destination = destination.Slice(minValues);
+        }
     }
+
+    private static string GetChoices(bool lower, bool upper, bool numeric, bool special) =>
+        (lower, upper, numeric, special) switch
+        {
+            (true, true, true, true) => LowerCaseChars + UpperCaseChars + NumericChars + SpecialChars,
```

Comment:
We don't need a table here, just 4 ifs ?

Reply:
Is lower allocations always strictly worth adding more constants? How often do we generate passwords? 
(Ofc not suggesting it actually matters what approach you use here)

Reply:
There comes a point where you can't "unsee" these allocations. Following the "4 ifs" approach would look something like this:

```C#
string result = null;
if (lower)
{
    result += LowerCaseChars;
}
if (upper)
{
    result += UpperCaseChars;
}
if (numeric)
{
    result += NumericChars;
}
if (special)
{
    result += SpecialChars;
}

if (result is null)
{
    throw new ArgumentException("At least one character type must be enabled.");
}

return result;
```

By default all bools are "true", so we'd allocate 3 throw away strings by default every time. I'm pretty sure someone would see this code and submit a PR to optimize it later.